
<!DOCTYPE html><html>
<head>
<meta charset="windows-1252">
<title>NGSANE project card</title>
<link rel="shortcut icon" href="includes/images/favicon.ico">
<link rel="apple-touch-icon" sizes="57x57" href="includes/images/apple-touch-icon-57x57.png">
<link rel="apple-touch-icon" sizes="114x114" href="includes/images/apple-touch-icon-114x114.png">
<link rel="apple-touch-icon" sizes="72x72" href="includes/images/apple-touch-icon-72x72.png">
<link rel="apple-touch-icon" sizes="144x144" href="includes/images/apple-touch-icon-144x144.png">
<link rel="apple-touch-icon" sizes="60x60" href="includes/images/apple-touch-icon-60x60.png">
<link rel="apple-touch-icon" sizes="120x120" href="includes/images/apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" sizes="76x76" href="includes/images/apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" sizes="152x152" href="includes/images/apple-touch-icon-152x152.png">
<link rel="icon" type="image/png" href="includes/images/favicon-196x196.png" sizes="196x196">
<link rel="icon" type="image/png" href="includes/images/favicon-160x160.png" sizes="160x160">
<link rel="icon" type="image/png" href="includes/images/favicon-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="includes/images/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="includes/images/favicon-32x32.png" sizes="32x32">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="msapplication-TileImage" content="includes/images/mstile-144x144.png">
<meta name="msapplication-config" content="includes/images/browserconfig.xml">
<script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script type="text/javascript">
if (!window.jQuery) {
    document.write("<script src=includes/js/jquery.js><\/script>");
}
</script>
<script type="text/javascript" src="http://cdn.datatables.net/1.10.2/js/jquery.dataTables.min.js"></script>
<script type="text/javascript">
if (!$.fn.dataTableExt) { 
    document.write("<script src=includes/js/jquery.dataTables.min.js><\/script>");
}
</script>
<link rel="stylesheet" type="text/css" href="http://cdn.datatables.net/1.10.2/css/jquery.dataTables.css">
<style>
body {
	font: 12px/19px "Lucida Sans", "Lucida Grande", "Lucida Sans Unicode",
		Verdana, sans-serif;
}

div.tabContent { margin: 0.5em; }
div.tabContent_hide { display: none; }
div#right{
	float:right;
	text-align:right;
	width:385px;
	padding:2px 12px 2px 12px;
}

h2 {
	display: block;
	font-size: 1.17em;
	-webkit-margin-before: 1em;
	-webkit-margin-after: 1em;
	-webkit-margin-start: 0px;
	-webkit-margin-end: 0px;
	font-weight: bold;
}

h4 {
	display: block;
	-webkit-margin-before: 1em;
    -webkit-margin-after: 0em;
	-webkit-margin-start: 0px;
	-webkit-margin-end: 0px;
}

.panel h2 {
	background-color: #999;
	border-radius: 10px;
	color: white;
	font-size: 100%;
	font-weight: normal;
	font-variant:small-caps;
	padding: 2px 10px;
	margin: 0px;
}

.panel h3 {
    border-top: 1px solid #69c;
    border-bottom: 1px dotted #69c;
	background-color: #efecff;
	color: #17212e;
	font-size: 100%;
	font-weight: normal;
	padding: 2px 10px;
	margin: 0 0 6 0;
}

.panel h3.overall {
	border-top: 3px double #69c;
	text-transform: small_caps;
}

div.panel h2,div.panel a h2.sub {
	background: #888;
	background-image: -webkit-gradient(linear, left top, left bottom, from(#8a9aa6),
		to(#17212e));
	background-image: -webkit-linear-gradient(top, #8a9aa6, #17212e);
	background-image: -moz-linear-gradient(top, #8a9aa6, #17212e);
	background-image: -ms-linear-gradient(top, #8a9aa6, #17212e);
	background-image: -o-linear-gradient(top, #8a9aa6, #17212e);
	background-image: linear-gradient(top, #8a9aa6, #17212e);
}

.panel#quicklinks h3{
	background-color: #8b5;
	color: white;
	text-transform: small_caps;
}

div.panel#quicklinks h2{
	background: #888;
	background-image: -webkit-gradient(linear, left top, left bottom, from(#8a9aa6),
		to(#17212e));
	background-image: -webkit-linear-gradient(top, #8a9aa6, #17212e);
	background-image: -moz-linear-gradient(top, #8a9aa6, #17212e);
	background-image: -ms-linear-gradient(top, #8a9aa6, #17212e);
	background-image: -o-linear-gradient(top, #8a9aa6, #17212e);
	background-image: linear-gradient(top, #8a9aa6, #17212e);
}

div.panel h2.sub {
	margin-left: 2px;
	margin-top: 2px;
	border-radius: 10px 10px 0 0;
	border: 1px solid #666;
	border-top-color: #EEE;
	border-left-color: #DDD;
	border-bottom: none;
	display: inline-block;
	vertical-align: top;
	width: 165px;
	color: #dedede;
	cursor: pointer;
}

div.headbagb {
	width: 100%;
	border-radius: 10px 10px 0 0;
	background: #555;
	background-image: -webkit-gradient(linear, left top, left bottom, from(#8a9aa6),
		to(#17212e));
	background-image: -webkit-linear-gradient(top, #8a9aa6, #17212e);
	background-image: -moz-linear-gradient(top, #8a9aa6, #17212e);
	background-image: -ms-linear-gradient(top, #8a9aa6, #17212e);
	background-image: -o-linear-gradient(top, #8a9aa6, #17212e);
	background-image: linear-gradient(top, #8a9aa6, #17212e);
}

div.box{
    padding: 5px;
}

div.scroll{
    overflow:scroll; 
    height:300px;
}

div.panel {
	background: #fafaff;
	border: 1px solid #999;
	border-radius: 12px;
	padding: 2px;
	margin-bottom: 25px;
	min-width:1170px;
}

#controls div,#gallery div {
	margin: 2px 0;
	width: 240px;
}

a {
	display: inline-block;
	outline: 0;
	color: #333;
}

#quicklinks a {
	padding: 2px;
	margin: 2px;
	transition-duration: 0.25s;
	transition-property: transform;
	transform: scale(1);
}

#quicklinks a:hover {
	background: #17212e;
	text-decoration: none;
	color: #fff;
	border-radius: 10px;
	transform: scale(1.05);
}

div.panel a.selected h2.sub {
	background: #fafaff;
	color: #444;
}

div.panel span.counter {
    background: none repeat scroll 0 0 #444;
    border-radius: 7px 7px 7px 7px;
    color: #333;
    display: inline;
    font-size: 90%;
    font-weight:bold;
    letter-spacing: 0;
    margin-left: 6px;
    text-transform: none;
}

div.panel span.counter span.passed{
    background: none repeat scroll 0 0 #dfd;
    border-radius: 7px 0px 0px 7px;
    padding: 0px 5px;
    color:#060;
    border: 1px solid #0a0;
}

div.panel span.counter span.failed{
    background: none repeat scroll 0 0 #866;
    border-radius: 0px 7px 7px 0px;
    padding: 0px 5px;
    color:#eee;
    border: 1px solid #a00
}

div.panel span.counter span.nofailed{
    background: none repeat scroll 0 0 #ddd;
    border-radius: 0px 7px 7px 0px;
    padding: 0px 5px;
    color:#666;
    border: 1px solid #444
}

div.panel span.counter span.errors{
    background: none repeat scroll 0 0 #866;
    border-radius: 7px 7px 7px 7px;
    padding: 0px 5px;
    color:#eee;
    border: 1px solid #a00;
}

div.panel span.counter span.neutral{
    background: none repeat scroll 0 0 #ddd;
    border-radius: 7px 7px 7px 7px;
    padding: 0px 5px;
    color:#666;
    border: 1px solid #444;
}

div.wrapper {
	margin: auto; 
	width: 100%;
	padding-top: 5px;
}

div.display {
	overflow-x: scroll;
	overflow-y: hidden;
	margin-bottom: 10px;
}

div.hidden {
    display:none;
}

p {
	color: blue;
}

.dataTables_filter{
    height:30px;
}

table.dataTable.display {
	font-size: 11px;
	text-align: left;
	border-collapse: collapse;
	border: 1px dotted #69c;
}

table.dataTable.display tbody td.dataTables_empty{
    text-align: center;
    font-style: italic;
}

table.dataTable.display thead .sorting, table.dataTable.display thead .sorting_asc, table.dataTable.display thead .sorting_desc{
    padding-right:15px;
}

table,caption,tbody,tfoot,thead,tr,th {
	font-size: 100%;
	vertical-align: middle;
	margin: 0;
	padding: 0;
	outline: 0;
	border: 0;
	background: transparent;
}

table.dataTable.display tbody tr.odd, table.dataTable.display tbody tr.odd, 
table.dataTable.display tbody tr.even>.sorting_1, table.dataTable.order-column.stripe tbody tr.even>.sorting_1,
table.dataTable.display tbody tr.even>.sorting_2, table.dataTable.order-column.stripe tbody tr.even>.sorting_2,
table.dataTable.display tbody tr.even>.sorting_3, table.dataTable.order-column.stripe tbody tr.even>.sorting_3{
    background: #efecff;
}
table.dataTable.display tbody tr.odd>.sorting_1, table.dataTable.order-column.stripe tbody tr.odd>.sorting_1,
table.dataTable.display tbody tr.odd>.sorting_2, table.dataTable.order-column.stripe tbody tr.odd>.sorting_2,
table.dataTable.display tbody tr.odd>.sorting_3, table.dataTable.order-column.stripe tbody tr.odd>.sorting_3{
    background: #dfdcef;
}
table.dataTable.display tbody tr.even, table.dataTable.display tbody tr.even{
    background: #f5f3ff;
}

table.dataTable.display th {
	font-weight: bold;
	color: #17212e;
	border-bottom: 1px dashed #69c;
	padding: 12px 5px;
	text-align:center;
}
table.dataTable tfoot td{
	border-top: 1px dashed #69c;
}

table.dataTable.display thead th div.DataTables_sort_wrapper {
  padding: 0
}

table.dataTable.display thead th {
  padding: 3px
}

table.dataTable.display th div{
	width: 100px;
}

table.dataTable.display tbody td, table.dataTable tfoot td {
	color: #555e66;
	padding: 3px 5px;
	text-align:right;
	white-space: nowrap;
}
table.dataTable tfoot td{
    font-style: italic;       
}

table.dataTable.display td.left, table.dataTable.display  th.left{
	text-align:left;
	width: auto;
}

.top{
	vertical-align:text-top
}

table.dataTable.display tfoot {
	border-top: 1px dashed #69c;
}

div.library {
	background: #fff;
	padding: 10px;
	max-width: 1200px;
}

pre {
    padding: 0 10px 0 10px;
}

hr {
	border: 0;
	border-top: 1px solid #ccc;
}
hr.spacer {
    padding-top:10px;
	border: 0;
	border-bottom: 1px dotted #69c;
}
span {
    color:#666;
    font-size:10px;
}

img.border {
    border: 1px solid #8a9aa6;
}

img {
    border: 0px;
    vertical-align: middle;
}

div.inset_image {
    float:left;
    text-align: center; 
    margin:2px; 
    padding:2px;
    border-width:1px;
    border-style:dotted;
}

</style>
</head>
<body>
<script type="text/javascript">
//<![CDATA[
    var datatable_array=[];

function getHash( url ) {
    var hashPos = url.lastIndexOf ( '#' );
    return url.substring( hashPos + 1 );
}

// format number with comma
function formatNumber(n) {
    // only attempt conversion if n is numberic
    if (!isNaN(parseFloat(n))){
        var t = Math.abs(parseFloat(n));
        if (t == 0)
            return n;
        else if (t < 0.01)
            return t.toExponential();
        else {
            // treat pre and post differently
            var n= (Math.round(t * 100) / 100).toString().split(".");
            // add comma to decimal positions only
            n[0] = n[0].replace(/\B(?=(\d{3})+(?!\d))/g, ",");
            //Combine both sections again
            return n.join(".");
        }
    } else
        return n;
}

Math.max = function() {
    var ary, m;
    ary = Array.prototype.slice.call(arguments);
    numA = function(a, b){return (b-a);};
    ary.sort(numA);
    m = ary[0];
    if (!isNaN(m) && isFinite(m))
        return m;
    return "";
};

Math.min = function() {
    var ary, m;
    ary = Array.prototype.slice.call(arguments);
    numA = function(a, b){return (a-b);};
    ary.sort(numA);
    m = ary[0];
    if (!isNaN(m) && isFinite(m))
        return m;
    return "";
};

Math.median = function() {
    var ary, numA, i, m;
    ary = Array.prototype.slice.call(arguments);
    numA = function(a, b){return (a-b);};
    ary.sort(numA);
    m = ary[Math.floor(ary.length/2)];
    if (!isNaN(m) && isFinite(m))
        return m;
    return "";
}


var isArray = function (obj) {
	return Object.prototype.toString.call(obj) === "[object Array]";
};

Math.mean = function( ){
    var ary, i, m, sum=0;
    ary = Array.prototype.slice.call(arguments);
    for (i = ary.length-1; i >= 0; i--) {sum += ary[i];}
	m = (sum / ary.length );
    
    if (!isNaN(m) && isFinite(m))
        return m;
    return "";

}

Math.variance = function( ){
    var ary, i, mean, v=0, sum=0;
    ary = Array.prototype.slice.call(arguments);
    for (i = ary.length-1; i >= 0; i--) {sum += ary[i];}
	mean = (sum / ary.length );
    console.log(ary)
    console.log(mean)
    for (i = ary.length-1; i >= 0; i--) {
		v += Math.pow( (ary[ i ] - mean), 2 );
	}
	v /= ary.length;
    if (!isNaN(v) && isFinite(v))
        return v;
    return "";
	
}

Math.std = function( ){
    var ary, avg, i, sddev, v=0, sum=0 ;
    ary = Array.prototype.slice.call(arguments);
    for (i = ary.length-1; i >= 0; i--) {sum += ary[i];}
	avg = (sum / ary.length );
    i = ary.length;
	while( i-- ){
		v += Math.pow( (ary[ i ] - avg), 2 );
	}
	sddev = Math.sqrt(v / ary.length);
    if (!isNaN(sddev) && isFinite(sddev))
        return sddev;
    return "";
	
}

//]]>
</script>
<div id="center">
<div class=panel id=quicklinks><h2>Quicklinks</h2><div>

<a href='#fastQC_panel' class='quicklinks'>fastQC</a>|<a href='#bowtie2_panel' class='quicklinks'>bowtie2</a>|<a href='#bigwig_panel' class='quicklinks'>bigwig</a>|<a href='#peakranger_panel' class='quicklinks'>peakranger</a>|<a href='#homerchipseq_panel' class='quicklinks'>homerchipseq</a>|<a href='#macs2_panel' class='quicklinks'>macs2</a>|<a href='#memechip_panel' class='quicklinks'>memechip</a>|<a href='#References_panel' class='quicklinks'>References</a>

<div id ="right"><label>Aggregate: <input id="showAggregation" type="checkbox" /></label> <label>Global Filter: <input type="search" id="search" /></label></div>
</div><!-- Links -->
</div><!-- panel -->
<div class='panel' id='fastQC_panel'><div class='tabs headbagb' id='fastQC_panel_tabs'><a name='fastQC'></a>
            <a class='selected' href='#fastQC_results'><h2 id='fastQC_h_results' class='sub'>FastQC</h2></a>
            <a href='#fastQC_checklist'><h2 id='fastQC_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='fastQC_counter_checkpoints_passed'></span><span class='failed' id='fastQC_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#fastQC_notes'><h2 id='fastQC_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='fastQC_counter_notes'></span></span></span></h2></a>
            <a href='#fastQC_errors'><h2 id='fastQC_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='fastQC_counter_errors'></span></span></h2></a>
            <a href='#fastQC_logfiles'><h2 id='fastQC_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
</div>
<div class='tabContent_hide' id='DC_fastQC_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/fastQC 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 3/3 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/3 have # QCVARIABLES
QC_PASS .. 0/3 have Resource temporarily unavailable
>>>>>>>>>> CheckPoints 
QC_PASS .. 3/3 have programs
QC_PASS .. 3/3 have parameters
QC_PASS .. 3/3 have recall files from tape
QC_PASS .. 3/3 have fastqc read 1
QC_PASS .. 3/3 have fastqc read 2

</pre></div></div>
<div class='tabContent_hide' id='DC_fastQC_notes'><div class='box scroll'><pre>

qout/fastQC/ChIPseq_Histone_H3k9me3_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/fastQC/job.20140910.log
[NOTE] set java parameters
[NOTE] Single-end pair library detected. No second read to process.

qout/fastQC/ChIPseq_input_Input_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/fastQC/job.20140910.log
[NOTE] set java parameters
[NOTE] Single-end pair library detected. No second read to process.

qout/fastQC/ChIPseq_TF_CTCF_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/fastQC/job.20140910.log
[NOTE] set java parameters
[NOTE] Single-end pair library detected. No second read to process.
</pre></div></div>
<div class='tabContent_hide' id='DC_fastQC_errors'><div class='box scroll'><pre>

qout/fastQC/ChIPseq_Histone_H3k9me3_chr16.out
-- all good, no errors

qout/fastQC/ChIPseq_input_Input_chr16.out
-- all good, no errors

qout/fastQC/ChIPseq_TF_CTCF_chr16.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_fastQC_logfiles'><div><div class='box scroll'>
<a href='../qout/fastQC/ChIPseq_Histone_H3k9me3_chr16.out'>ChIPseq_Histone_H3k9me3_chr16.out</a><br/>
<a href='../qout/fastQC/ChIPseq_input_Input_chr16.out'>ChIPseq_input_Input_chr16.out</a><br/>
<a href='../qout/fastQC/ChIPseq_TF_CTCF_chr16.out'>ChIPseq_TF_CTCF_chr16.out</a><br/>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#fastQC_counter_notes').text('9');
            $('#fastQC_counter_errors').text('0');
            $('#fastQC_counter_checkpoints_passed').text('8');
            $('#fastQC_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#fastQC_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#fastQC_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#fastQC_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#fastQC_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_fastQC_results'><div>
<script type="text/javascript">
//<![CDATA[
fastQC_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'', 'bSortable':false , 'sWidth':'20px'} ,{'sTitle':'', 'mRender': function(data){return formatNumber(data);} , 'bSortable':false , 'sWidth':'15px'} ,{'sTitle':'', 'bSortable':false , 'sWidth':'20px'} ,{'sTitle':'', 'mRender': function(data){return formatNumber(data);} , 'bSortable':false , 'sWidth':'15px'} ,{'sTitle':'', 'bSortable':false , 'sWidth':'20px'} ,{'sTitle':'', 'mRender': function(data){return formatNumber(data);} , 'bSortable':false , 'sWidth':'15px'} ,{'sTitle':'Encoding', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Library size', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Length', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'80px'} ,{'sTitle':'%GC', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'', 'bSortable':false , 'sWidth':'90px'} ,{'sTitle':'', 'bSortable':false , 'sWidth':'90px'} ,{'sTitle':'', 'bSortable':false , 'sWidth':'90px'} ,{'sTitle':'', 'bSortable':false , 'sWidth':'90px'}
], "aaData": [
['ChIPseq_TF' , '<a target="detail" href="../ChIPseq_TF/fastQC/CTCF_chr16_R1.html">CTCF_chr16_R1</a>' , '<a target="img" href="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Icons/tick.png"><img 0 src="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Icons/tick.png" height=15/></a>' , 8.00 , '<a target="img" href="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Icons/warning.png"><img 0 src="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Icons/warning.png" height=15/></a>' , 3.00 , '<a target="img" href="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Icons/error.png"><img 0 src="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Icons/error.png" height=15/></a>' , 1.00 , 'Sanger / Illumina 1.9' , 7482.00 , 36.00 , 51.00 , '<a target="img" href="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Images/per_base_quality.png"><img class=\"border\" src="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Images/per_base_quality.png" height=60/></a>' , '<a target="img" href="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Images/duplication_levels.png"><img class=\"border\" src="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Images/duplication_levels.png" height=60/></a>' , '<a target="img" href="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Images/per_sequence_gc_content.png"><img class=\"border\" src="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Images/per_sequence_gc_content.png" height=60/></a>' , '<a target="img" href="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Images/per_base_sequence_content.png"><img class=\"border\" src="../ChIPseq_TF/fastQC/CTCF_chr16_R1/Images/per_base_sequence_content.png" height=60/></a>' ]
,['ChIPseq_Histone' , '<a target="detail" href="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1.html">H3k9me3_chr16_R1</a>' , '<a target="img" href="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Icons/tick.png"><img 0 src="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Icons/tick.png" height=15/></a>' , 9.00 , '<a target="img" href="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Icons/warning.png"><img 0 src="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Icons/warning.png" height=15/></a>' , 2.00 , '<a target="img" href="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Icons/error.png"><img 0 src="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Icons/error.png" height=15/></a>' , 1.00 , 'Sanger / Illumina 1.9' , 1942.00 , 36.00 , 45.00 , '<a target="img" href="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Images/per_base_quality.png"><img class=\"border\" src="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Images/per_base_quality.png" height=60/></a>' , '<a target="img" href="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Images/duplication_levels.png"><img class=\"border\" src="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Images/duplication_levels.png" height=60/></a>' , '<a target="img" href="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Images/per_sequence_gc_content.png"><img class=\"border\" src="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Images/per_sequence_gc_content.png" height=60/></a>' , '<a target="img" href="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Images/per_base_sequence_content.png"><img class=\"border\" src="../ChIPseq_Histone/fastQC/H3k9me3_chr16_R1/Images/per_base_sequence_content.png" height=60/></a>' ]
,['ChIPseq_input' , '<a target="detail" href="../ChIPseq_input/fastQC/Input_chr16_R1.html">Input_chr16_R1</a>' , '<a target="img" href="../ChIPseq_input/fastQC/Input_chr16_R1/Icons/tick.png"><img 0 src="../ChIPseq_input/fastQC/Input_chr16_R1/Icons/tick.png" height=15/></a>' , 11.00 , '<a target="img" href="../ChIPseq_input/fastQC/Input_chr16_R1/Icons/warning.png"><img 0 src="../ChIPseq_input/fastQC/Input_chr16_R1/Icons/warning.png" height=15/></a>' , 0.00 , '<a target="img" href="../ChIPseq_input/fastQC/Input_chr16_R1/Icons/error.png"><img 0 src="../ChIPseq_input/fastQC/Input_chr16_R1/Icons/error.png" height=15/></a>' , 1.00 , 'Sanger / Illumina 1.9' , 4114.00 , 36.00 , 48.00 , '<a target="img" href="../ChIPseq_input/fastQC/Input_chr16_R1/Images/per_base_quality.png"><img class=\"border\" src="../ChIPseq_input/fastQC/Input_chr16_R1/Images/per_base_quality.png" height=60/></a>' , '<a target="img" href="../ChIPseq_input/fastQC/Input_chr16_R1/Images/duplication_levels.png"><img class=\"border\" src="../ChIPseq_input/fastQC/Input_chr16_R1/Images/duplication_levels.png" height=60/></a>' , '<a target="img" href="../ChIPseq_input/fastQC/Input_chr16_R1/Images/per_sequence_gc_content.png"><img class=\"border\" src="../ChIPseq_input/fastQC/Input_chr16_R1/Images/per_sequence_gc_content.png" height=60/></a>' , '<a target="img" href="../ChIPseq_input/fastQC/Input_chr16_R1/Images/per_base_sequence_content.png"><img class=\"border\" src="../ChIPseq_input/fastQC/Input_chr16_R1/Images/per_base_sequence_content.png" height=60/></a>' ]
        ]};
        datatable_array.push({"json":"fastQC_Table_1_json","html":"fastQC_id_Table_1"});
        //]]>
        </script>
    <table id="fastQC_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel' id='bowtie2_panel'><div class='tabs headbagb' id='bowtie2_panel_tabs'><a name='bowtie2'></a>
            <a class='selected' href='#bowtie2_results'><h2 id='bowtie2_h_results' class='sub'>Bowtie2</h2></a>
            <a href='#bowtie2_checklist'><h2 id='bowtie2_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='bowtie2_counter_checkpoints_passed'></span><span class='failed' id='bowtie2_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#bowtie2_notes'><h2 id='bowtie2_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='bowtie2_counter_notes'></span></span></span></h2></a>
            <a href='#bowtie2_errors'><h2 id='bowtie2_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='bowtie2_counter_errors'></span></span></h2></a>
            <a href='#bowtie2_logfiles'><h2 id='bowtie2_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
<a href='#bowtie2_PRIMARY_RESULT_FILES'><h2 id='bowtie2_h_nrfiles' class='sub inactive' rel='files'>Primary result files</h2></a>
</div>
<div class='tabContent_hide' id='DC_bowtie2_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/bowtie2 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 3/3 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/3 have # QCVARIABLES
QC_PASS .. 0/3 have Resource temporarily unavailable
>>>>>>>>>> CheckPoints 
QC_PASS .. 3/3 have programs
QC_PASS .. 3/3 have parameters
QC_PASS .. 3/3 have recall files from tape
QC_PASS .. 3/3 have map with bowtie2
QC_PASS .. 3/3 have clean sam
QC_PASS .. 3/3 have mark duplicates
QC_PASS .. 3/3 have statistics
QC_PASS .. 3/3 have calculate inner distance
QC_PASS .. 3/3 have samstat
QC_PASS .. 3/3 have verify
QC_PASS .. 3/3 have cleanup

</pre></div></div>
<div class='tabContent_hide' id='DC_bowtie2_notes'><div class='box scroll'><pre>

qout/bowtie2/ChIPseq_Histone_H3k9me3_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bowtie2/job.20140910.log
[NOTE] set java parameters
[NOTE] EXPID CTCF; LIBRARY ChIPseq; PLATFORM Illumina
[NOTE] Sanger Phred33	(34, 66) fastq format detected
[NOTE] PASS check mapping: 1942 == 1942

qout/bowtie2/ChIPseq_input_Input_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bowtie2/job.20140910.log
[NOTE] set java parameters
[NOTE] EXPID CTCF; LIBRARY ChIPseq; PLATFORM Illumina
[NOTE] Sanger Phred33	(33, 67) fastq format detected
[NOTE] PASS check mapping: 4114 == 4114

qout/bowtie2/ChIPseq_TF_CTCF_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bowtie2/job.20140910.log
[NOTE] set java parameters
[NOTE] EXPID CTCF; LIBRARY ChIPseq; PLATFORM Illumina
[NOTE] Sanger Phred33	(33, 67) fastq format detected
[NOTE] PASS check mapping: 7482 == 7482
</pre></div></div>
<div class='tabContent_hide' id='DC_bowtie2_errors'><div class='box scroll'><pre>

qout/bowtie2/ChIPseq_Histone_H3k9me3_chr16.out
-- all good, no errors

qout/bowtie2/ChIPseq_input_Input_chr16.out
-- all good, no errors

qout/bowtie2/ChIPseq_TF_CTCF_chr16.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_bowtie2_logfiles'><div><div class='box scroll'>
<a href='../qout/bowtie2/ChIPseq_Histone_H3k9me3_chr16.out'>ChIPseq_Histone_H3k9me3_chr16.out</a><br/>
<a href='../qout/bowtie2/ChIPseq_input_Input_chr16.out'>ChIPseq_input_Input_chr16.out</a><br/>
<a href='../qout/bowtie2/ChIPseq_TF_CTCF_chr16.out'>ChIPseq_TF_CTCF_chr16.out</a><br/>
</div></div></div>
<div class='tabContent_hide' id='DC_bowtie2_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
<a href='../ChIPseq_TF/bowtie2/CTCF_chr16.asd.bam'>CTCF_chr16.asd.bam</a><br/>
<a href='../ChIPseq_Histone/bowtie2/H3k9me3_chr16.asd.bam'>H3k9me3_chr16.asd.bam</a><br/>
<a href='../ChIPseq_input/bowtie2/Input_chr16.asd.bam'>Input_chr16.asd.bam</a><br/>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#bowtie2_counter_notes').text('15');
            $('#bowtie2_counter_errors').text('0');
            $('#bowtie2_counter_checkpoints_passed').text('14');
            $('#bowtie2_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#bowtie2_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#bowtie2_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#bowtie2_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#bowtie2_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_bowtie2_results'><div>
<script type="text/javascript">
//<![CDATA[
bowtie2_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Total reads', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Qfail', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Duplicates', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'%', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'Mapped', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'%', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'Mapped pairs', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'%', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'Singletons', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
['ChIPseq_TF' , '<a target="detail" href="../ChIPseq_TF/bowtie2/CTCF_chr16.asd.bam.html">CTCF_chr16</a>' , 7482.00 , 0.00 , 1221.00 , 16.31 , 6198.00 , 82.83 , 0.00 , 0 , 0.00 ]
,['ChIPseq_Histone' , '<a target="detail" href="../ChIPseq_Histone/bowtie2/H3k9me3_chr16.asd.bam.html">H3k9me3_chr16</a>' , 1942.00 , 0.00 , 398.00 , 20.49 , 1935.00 , 99.63 , 0.00 , 0 , 0.00 ]
,['ChIPseq_input' , '<a target="detail" href="../ChIPseq_input/bowtie2/Input_chr16.asd.bam.html">Input_chr16</a>' , 4114.00 , 0.00 , 145.00 , 3.52 , 3999.00 , 97.20 , 0.00 , 0 , 0.00 ]
        ]};
        datatable_array.push({"json":"bowtie2_Table_1_json","html":"bowtie2_id_Table_1"});
        //]]>
        </script>
    <table id="bowtie2_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel' id='bigwig_panel'><div class='tabs headbagb' id='bigwig_panel_tabs'><a name='bigwig'></a>
            <a class='selected' href='#bigwig_results'><h2 id='bigwig_h_results' class='sub'>Bigwig</h2></a>
            <a href='#bigwig_checklist'><h2 id='bigwig_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='bigwig_counter_checkpoints_passed'></span><span class='failed' id='bigwig_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#bigwig_notes'><h2 id='bigwig_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='bigwig_counter_notes'></span></span></span></h2></a>
            <a href='#bigwig_errors'><h2 id='bigwig_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='bigwig_counter_errors'></span></span></h2></a>
            <a href='#bigwig_logfiles'><h2 id='bigwig_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
<a href='#bigwig_PRIMARY_RESULT_FILES'><h2 id='bigwig_h_nrfiles' class='sub inactive' rel='files'>Primary result files</h2></a>
</div>
<div class='tabContent_hide' id='DC_bigwig_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/bigwig 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 3/3 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/3 have # QCVARIABLES
QC_PASS .. 0/3 have Resource temporarily unavailable
>>>>>>>>>> CheckPoints 
QC_PASS .. 3/3 have programs
QC_PASS .. 3/3 have parameters
QC_PASS .. 3/3 have recall files from tape
QC_PASS .. 3/3 have generate bigwigs

</pre></div></div>
<div class='tabContent_hide' id='DC_bigwig_notes'><div class='box scroll'><pre>

qout/bigwig/ChIPseq_Histone_H3k9me3_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bigwig/job.20140910.log
[NOTE] Reference: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.fasta
[NOTE] Chromosome size: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.chrom.sizes
[NOTE] single-end library detected
[NOTE] fragment length: 300
[NOTE] generate bigwig with specified fragment length

qout/bigwig/ChIPseq_input_Input_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bigwig/job.20140910.log
[NOTE] Reference: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.fasta
[NOTE] Chromosome size: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.chrom.sizes
[NOTE] single-end library detected
[NOTE] fragment length: 300
[NOTE] generate bigwig with specified fragment length

qout/bigwig/ChIPseq_TF_CTCF_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bigwig/job.20140910.log
[NOTE] Reference: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.fasta
[NOTE] Chromosome size: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.chrom.sizes
[NOTE] single-end library detected
[NOTE] fragment length: 300
[NOTE] generate bigwig with specified fragment length
</pre></div></div>
<div class='tabContent_hide' id='DC_bigwig_errors'><div class='box scroll'><pre>

qout/bigwig/ChIPseq_Histone_H3k9me3_chr16.out
-- all good, no errors

qout/bigwig/ChIPseq_input_Input_chr16.out
-- all good, no errors

qout/bigwig/ChIPseq_TF_CTCF_chr16.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_bigwig_logfiles'><div><div class='box scroll'>
<a href='../qout/bigwig/ChIPseq_Histone_H3k9me3_chr16.out'>ChIPseq_Histone_H3k9me3_chr16.out</a><br/>
<a href='../qout/bigwig/ChIPseq_input_Input_chr16.out'>ChIPseq_input_Input_chr16.out</a><br/>
<a href='../qout/bigwig/ChIPseq_TF_CTCF_chr16.out'>ChIPseq_TF_CTCF_chr16.out</a><br/>
</div></div></div>
<div class='tabContent_hide' id='DC_bigwig_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
<a href='../ChIPseq_TF/bigwig/CTCF_chr16.bw'>CTCF_chr16.bw</a><br/>
<a href='../ChIPseq_Histone/bigwig/H3k9me3_chr16.bw'>H3k9me3_chr16.bw</a><br/>
<a href='../ChIPseq_input/bigwig/Input_chr16.bw'>Input_chr16.bw</a><br/>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#bigwig_counter_notes').text('18');
            $('#bigwig_counter_errors').text('0');
            $('#bigwig_counter_checkpoints_passed').text('7');
            $('#bigwig_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#bigwig_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#bigwig_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#bigwig_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#bigwig_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_bigwig_results'><div>
<script type="text/javascript">
//<![CDATA[
bigwig_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'library size', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'normalize to', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'scale factor', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
['ChIPseq_TF' , 'CTCF_chr16' , 4977.00 , 'NA' , 1.00 ]
,['ChIPseq_Histone' , 'H3k9me3_chr16' , 1537.00 , 'NA' , 1.00 ]
,['ChIPseq_input' , 'Input_chr16' , 3854.00 , 'NA' , 1.00 ]
        ]};
        datatable_array.push({"json":"bigwig_Table_1_json","html":"bigwig_id_Table_1"});
        //]]>
        </script>
    <table id="bigwig_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel' id='peakranger_panel'><div class='tabs headbagb' id='peakranger_panel_tabs'><a name='peakranger'></a>
            <a class='selected' href='#peakranger_results'><h2 id='peakranger_h_results' class='sub'>Peakranger</h2></a>
            <a href='#peakranger_checklist'><h2 id='peakranger_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='peakranger_counter_checkpoints_passed'></span><span class='failed' id='peakranger_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#peakranger_notes'><h2 id='peakranger_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='peakranger_counter_notes'></span></span></span></h2></a>
            <a href='#peakranger_errors'><h2 id='peakranger_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='peakranger_counter_errors'></span></span></h2></a>
            <a href='#peakranger_logfiles'><h2 id='peakranger_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
<a href='#peakranger_PRIMARY_RESULT_FILES'><h2 id='peakranger_h_nrfiles' class='sub inactive' rel='files'>Primary result files</h2></a>
</div>
<div class='tabContent_hide' id='DC_peakranger_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/peakranger 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 1/1 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/1 have # QCVARIABLES
QC_PASS .. 0/1 have Resource temporarily unavailable
>>>>>>>>>> CheckPoints 
QC_PASS .. 1/1 have programs
QC_PASS .. 1/1 have parameters
QC_PASS .. 1/1 have recall files from tape
QC_PASS .. 1/1 have assess data quality
QC_PASS .. 1/1 have peakranger
QC_PASS .. 1/1 have zip

</pre></div></div>
<div class='tabContent_hide' id='DC_peakranger_notes'><div class='box scroll'><pre>

qout/peakranger/ChIPseq_Histone_H3k9me3_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/peakranger/job.20140910.log
[NOTE] Chromosome size: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.chrom.sizes
[NOTE] calling broad peaks
</pre></div></div>
<div class='tabContent_hide' id='DC_peakranger_errors'><div class='box scroll'><pre>

qout/peakranger/ChIPseq_Histone_H3k9me3_chr16.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_peakranger_logfiles'><div><div class='box scroll'>
<a href='../qout/peakranger/ChIPseq_Histone_H3k9me3_chr16.out'>ChIPseq_Histone_H3k9me3_chr16.out</a><br/>
</div></div></div>
<div class='tabContent_hide' id='DC_peakranger_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
<a href='../ChIPseq_Histone/peakranger/H3k9me3_chr16_region.bed'>H3k9me3_chr16_region.bed</a><br/>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#peakranger_counter_notes').text('3');
            $('#peakranger_counter_errors').text('0');
            $('#peakranger_counter_checkpoints_passed').text('9');
            $('#peakranger_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#peakranger_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#peakranger_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#peakranger_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#peakranger_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_peakranger_results'><div>
<script type="text/javascript">
//<![CDATA[
peakranger_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Control', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Total reads', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Unique', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Unique %', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'80px'} ,{'sTitle':'Noise rate', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'100px'} ,{'sTitle':'Library complexity', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'100px'} ,{'sTitle':'Peaks', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Nucleotides covered', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
['ChIPseq_Histone' , 'H3k9me3_chr16' , 'Input_chr16' , 1935.00 , 1537.00 , 79.43 , 1.00 , 79.40 , 32.00 , 67250.00 ]
        ]};
        datatable_array.push({"json":"peakranger_Table_1_json","html":"peakranger_id_Table_1"});
        //]]>
        </script>
    <table id="peakranger_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel' id='homerchipseq_panel'><div class='tabs headbagb' id='homerchipseq_panel_tabs'><a name='homerchipseq'></a>
            <a class='selected' href='#homerchipseq_results'><h2 id='homerchipseq_h_results' class='sub'>Homer ChIP-seq</h2></a>
            <a href='#homerchipseq_checklist'><h2 id='homerchipseq_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='homerchipseq_counter_checkpoints_passed'></span><span class='failed' id='homerchipseq_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#homerchipseq_notes'><h2 id='homerchipseq_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='homerchipseq_counter_notes'></span></span></span></h2></a>
            <a href='#homerchipseq_errors'><h2 id='homerchipseq_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='homerchipseq_counter_errors'></span></span></h2></a>
            <a href='#homerchipseq_logfiles'><h2 id='homerchipseq_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
<a href='#homerchipseq_PRIMARY_RESULT_FILES'><h2 id='homerchipseq_h_nrfiles' class='sub inactive' rel='files'>Primary result files</h2></a>
</div>
<div class='tabContent_hide' id='DC_homerchipseq_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/chipseqHomer 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 1/1 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/1 have # QCVARIABLES
QC_PASS .. 0/1 have Resource temporarily unavailable
>>>>>>>>>> CheckPoints 
QC_PASS .. 1/1 have programs
QC_PASS .. 1/1 have parameters
QC_PASS .. 1/1 have recall files from tape
QC_PASS .. 1/1 have create tagdirectory
QC_PASS .. 1/1 have find peaks
QC_PASS .. 1/1 have summarize
QC_PASS .. 1/1 have cleanup

</pre></div></div>
<div class='tabContent_hide' id='DC_homerchipseq_notes'><div class='box scroll'><pre>

qout/homerchipseq/ChIPseq_TF_CTCF_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/homerchipseq/job.20140910.log
[NOTE] Chromosome size: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.chrom.sizes
</pre></div></div>
<div class='tabContent_hide' id='DC_homerchipseq_errors'><div class='box scroll'><pre>

qout/homerchipseq/ChIPseq_TF_CTCF_chr16.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_homerchipseq_logfiles'><div><div class='box scroll'>
<a href='../qout/homerchipseq/ChIPseq_TF_CTCF_chr16.out'>ChIPseq_TF_CTCF_chr16.out</a><br/>
</div></div></div>
<div class='tabContent_hide' id='DC_homerchipseq_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
<a href='../ChIPseq_TF/homerchipseq/CTCF_chr16-Input_chr16.bed'>CTCF_chr16-Input_chr16.bed</a><br/>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#homerchipseq_counter_notes').text('2');
            $('#homerchipseq_counter_errors').text('0');
            $('#homerchipseq_counter_checkpoints_passed').text('10');
            $('#homerchipseq_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#homerchipseq_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#homerchipseq_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#homerchipseq_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#homerchipseq_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_homerchipseq_results'><div>
<script type="text/javascript">
//<![CDATA[
homerchipseq_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Control', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Est. fragment length', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Autocorr. same strand', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Autocorr. diff. strand', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Autocorr. same/diff', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Peaks', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'100px'} ,{'sTitle':'Peak size', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'100px'} ,{'sTitle':'IP efficiency %', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
['ChIPseq_TF' , 'CTCF_chr16' , '' , 124.00 , 8.30 , 7.70 , 1.00 , 31.00 , 200.00 , 17.66 ]
        ]};
        datatable_array.push({"json":"homerchipseq_Table_1_json","html":"homerchipseq_id_Table_1"});
        //]]>
        </script>
    <table id="homerchipseq_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel' id='macs2_panel'><div class='tabs headbagb' id='macs2_panel_tabs'><a name='macs2'></a>
            <a class='selected' href='#macs2_results'><h2 id='macs2_h_results' class='sub'>Macs2</h2></a>
            <a href='#macs2_checklist'><h2 id='macs2_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='macs2_counter_checkpoints_passed'></span><span class='failed' id='macs2_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#macs2_notes'><h2 id='macs2_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='macs2_counter_notes'></span></span></span></h2></a>
            <a href='#macs2_errors'><h2 id='macs2_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='macs2_counter_errors'></span></span></h2></a>
            <a href='#macs2_logfiles'><h2 id='macs2_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
<a href='#macs2_PRIMARY_RESULT_FILES'><h2 id='macs2_h_nrfiles' class='sub inactive' rel='files'>Primary result files</h2></a>
</div>
<div class='tabContent_hide' id='DC_macs2_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/macs2 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 1/1 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/1 have # QCVARIABLES
QC_PASS .. 0/1 have Resource temporarily unavailable
>>>>>>>>>> CheckPoints 
QC_PASS .. 1/1 have programs
QC_PASS .. 1/1 have parameters
QC_PASS .. 1/1 have recall files from tape
QC_PASS .. 1/1 have macs 2 - predictd 
QC_PASS .. 1/1 have macs 2 - call peaks 
QC_PASS .. 1/1 have macs 2 - refine peaks 

</pre></div></div>
<div class='tabContent_hide' id='DC_macs2_notes'><div class='box scroll'><pre>

qout/macs2/ChIPseq_TF_CTCF_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/macs2/job.20140910.log
[NOTE] Reference: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.fasta
[NOTE] Chromosome size: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.chrom.sizes
[NOTE] CHIPINPUT /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/ChIPseq_input/bowtie2/Input_chr16.asd.bam
[NOTE] Skip modeling
[NOTE] convert narrowpeaks to 6 bed file for refining
[NOTE] create bigbed from peaks
</pre></div></div>
<div class='tabContent_hide' id='DC_macs2_errors'><div class='box scroll'><pre>

qout/macs2/ChIPseq_TF_CTCF_chr16.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_macs2_logfiles'><div><div class='box scroll'>
<a href='../qout/macs2/ChIPseq_TF_CTCF_chr16.out'>ChIPseq_TF_CTCF_chr16.out</a><br/>
</div></div></div>
<div class='tabContent_hide' id='DC_macs2_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
<a href='../ChIPseq_TF/macs2/CTCF_chr16_peaks.bed'>CTCF_chr16_peaks.bed</a><br/>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#macs2_counter_notes').text('7');
            $('#macs2_counter_errors').text('0');
            $('#macs2_counter_checkpoints_passed').text('9');
            $('#macs2_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#macs2_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#macs2_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#macs2_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#macs2_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_macs2_results'><div>
<script type="text/javascript">
//<![CDATA[
macs2_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Control', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'IP reads (filtered)', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Control reads (filtered)', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Paired peaks', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'100px'} ,{'sTitle':'Fragment size', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'100px'} ,{'sTitle':'Refined summits', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Nucleotides covered', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'', 'bSortable':false , 'sWidth':'60px'} ,{'sTitle':'', 'bSortable':false , 'sWidth':'60px'}
], "aaData": [
['ChIPseq_TF' , 'CTCF_chr16' , 'Input_chr16' , 4977.00 , 3854.00 , '' , 200.00 , 23.00 , 6477.00 , '' , '' ]
        ]};
        datatable_array.push({"json":"macs2_Table_1_json","html":"macs2_id_Table_1"});
        //]]>
        </script>
    <table id="macs2_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel' id='memechip_panel'><div class='tabs headbagb' id='memechip_panel_tabs'><a name='memechip'></a>
            <a class='selected' href='#memechip_results'><h2 id='memechip_h_results' class='sub'>MemeChip</h2></a>
            <a href='#memechip_checklist'><h2 id='memechip_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='memechip_counter_checkpoints_passed'></span><span class='failed' id='memechip_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#memechip_notes'><h2 id='memechip_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='memechip_counter_notes'></span></span></span></h2></a>
            <a href='#memechip_errors'><h2 id='memechip_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='memechip_counter_errors'></span></span></h2></a>
            <a href='#memechip_logfiles'><h2 id='memechip_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
</div>
<div class='tabContent_hide' id='DC_memechip_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/memechip 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 1/1 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/1 have # QCVARIABLES
QC_PASS .. 0/1 have Resource temporarily unavailable
>>>>>>>>>> CheckPoints 
QC_PASS .. 1/1 have programs
QC_PASS .. 1/1 have parameters
QC_PASS .. 1/1 have recall files from tape
QC_PASS .. 1/1 have get sequence data
QC_PASS .. 1/1 have create background model
QC_PASS .. 1/1 have meme-chip
QC_PASS .. 1/1 have classify bound regions
QC_PASS .. 1/1 have cleanup

</pre></div></div>
<div class='tabContent_hide' id='DC_memechip_notes'><div class='box scroll'><pre>

qout/memechip/ChIPseq_TF_CTCF_chr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/memechip/job.20140910.log
[NOTE] Reference: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.fasta
[NOTE] Chromosome size: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/chr16.chrom.sizes
[NOTE] extend bed regions: 
[NOTE] create background from peak regions
</pre></div></div>
<div class='tabContent_hide' id='DC_memechip_errors'><div class='box scroll'><pre>

qout/memechip/ChIPseq_TF_CTCF_chr16.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_memechip_logfiles'><div><div class='box scroll'>
<a href='../qout/memechip/ChIPseq_TF_CTCF_chr16.out'>ChIPseq_TF_CTCF_chr16.out</a><br/>
</div></div></div>
<div class='tabContent_hide' id='DC_memechip_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#memechip_counter_notes').text('5');
            $('#memechip_counter_errors').text('0');
            $('#memechip_counter_checkpoints_passed').text('11');
            $('#memechip_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#memechip_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#memechip_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#memechip_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#memechip_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_memechip_results'><div>
<script type="text/javascript">
//<![CDATA[
memechip_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Query consensus', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'E-value', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'80px'} ,{'sTitle':'Q-value', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'80px'} ,{'sTitle':'Similar to', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Peaks', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'With strong motif', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'', 'bSortable':false , 'sWidth':'150px'}
], "aaData": [
['ChIPseq_TF' , '<a target="detail" href="../ChIPseq_TF/memechip/CTCF_chr16/index.html">CTCF_chr16</a>' , 'TGCCACCAGGGGGCG' , '8.9e-015' , '1.6639e-05' , 'CTCF_full' , 24.00 , 23.00 , '<a target="img" href="../ChIPseq_TF/memechip/CTCF_chr16_topmememotif.png"><img class=\"border\" src="../ChIPseq_TF/memechip/CTCF_chr16_topmememotif.png" height=60/></a>' ]
        ]};
        datatable_array.push({"json":"memechip_Table_1_json","html":"memechip_id_Table_1"});
        //]]>
        </script>
    <table id="memechip_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel'><div class='headbagb'><a name='References_panel'><h2 class='sub'>References</h2></a></div>
<h3>Please cite the following publications/resources</h3>
<table class="data"><thead><tr><th><div style="width:10px"></div></th><th><div></div></th></tr></thead><tbody>
<tr class='citation'><td class='top'>[1]</td><td class='left' >Bioinformatics. 2009 Aug 15;25(16):2078-9. doi: 10.1093/bioinformatics/btp352. Epub 2009 Jun 8. 'The Sequence Alignment/Map format and SAMtools.'; Li H, Handsaker B, Wysoker A, Fennell T, Ruan J, Homer N, Marth G, Abecasis G, Durbin R; 1000 Genome Project Data Processing Subgroup.</td></tr>
<tr class='citation'><td class='top'>[2]</td><td class='left' >Bioinformatics. 2010 Mar 15;26(6):841-2. doi: 10.1093/bioinformatics/btq033. Epub 2010 Jan 28. 'BEDTools: a flexible suite of utilities for comparing genomic features.; Quinlan AR, Hall IM.</td></tr>
<tr class='citation'><td class='top'>[3]</td><td class='left' >Bioinformatics. 2011 Jan 1;27(1):130-1. doi: 10.1093/bioinformatics/btq614. Epub 2010 Nov 18. 'SAMStat: monitoring biases in next generation sequencing data.'; Lassmann T, Hayashizaki Y, Daub CO.</td></tr>
<tr class='citation'><td class='top'>[4]</td><td class='left' >Bioinformatics 2014; doi: 10.1093/bioinformatics/btu036 'NGSANE: A Lightweight Production Informatics Framework for High Throughput Data Analysis; Buske FA, French HJ, Smith MA, Clark SJ, Bauer DC</td></tr>
<tr class='citation'><td class='top'>[5]</td><td class='left' >BMC Bioinformatics. 2011 May 9;12:139. doi: 10.1186/1471-2105-12-139. 'PeakRanger: a cloud-enabled peak caller for ChIP-seq data.'; Feng X, Grossman R, Stein L.</td></tr>
<tr class='citation'><td class='top'>[6]</td><td class='left' >Brief Bioinform. 2013 Mar;14(2):144-61. doi: 10.1093/bib/bbs038. Epub 2012 Aug 20. 'The UCSC genome browser and associated tools.'; Kuhn RM, Haussler D, Kent WJ.</td></tr>
<tr class='citation'><td class='top'>[7]</td><td class='left' >Brief Bioinform. 2013 Mar;14(2):178-92. doi: 10.1093/bib/bbs017. Epub 2012 Apr 19. 'Integrative Genomics Viewer (IGV): high-performance genomics data visualization and exploration.'; Thorvaldsdóttir H, Robinson JT, Mesirov JP.</td></tr>
<tr class='citation'><td class='top'>[8]</td><td class='left' >Genome Biol. 2008;9(9):R137. doi: 10.1186/gb-2008-9-9-r137. Epub 2008 Sep 17. 'Model-based analysis of ChIP-Seq (MACS).'; Zhang Y, Liu T, Meyer CA, Eeckhoute J, Johnson DS, Bernstein BE, Nusbaum C, Myers RM, Brown M, Li W, Liu XS.</td></tr>
<tr class='citation'><td class='top'>[9]</td><td class='left' >Genome Res. 2002 Apr;12(4):656-64. 'BLAT--the BLAST-like alignment tool.'; Kent WJ.</td></tr>
<tr class='citation'><td class='top'>[10]</td><td class='left' >http://picard.sourceforge.net.</td></tr>
<tr class='citation'><td class='top'>[11]</td><td class='left' >http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</td></tr>
<tr class='citation'><td class='top'>[12]</td><td class='left' >Mol Cell. 2010 May 28;38(4):576-89. doi: 10.1016/j.molcel.2010.05.004. 'Simple combinations of lineage-determining transcription factors prime cis-regulatory elements required for macrophage and B cell identities.'; Heinz S, Benner C, Spann N, Bertolino E, Lin YC, Laslo P, Cheng JX, Murre C, Singh H, Glass CK.</td></tr>
<tr class='citation'><td class='top'>[13]</td><td class='left' >Nat Methods. 2012 Mar 4;9(4):357-9. doi: 10.1038/nmeth.1923. 'Fast gapped-read alignment with Bowtie 2.';Langmead B, Salzberg SL.</td></tr>
<tr class='citation'><td class='top'>[14]</td><td class='left' >Nucleic Acids Res. 1990 Oct 25;18(20):6097-100.; 'Sequence logos: a new way to display consensus sequences.'; Schneider TD, Stephens RM.</td></tr>
<tr class='citation'><td class='top'>[15]</td><td class='left' >Nucleic Acids Res. 2009 Jul;37(Web Server issue):W202-8. doi: 10.1093/nar/gkp335. Epub 2009 May 20. 'MEME SUITE: tools for motif discovery and searching.';Bailey TL, Boden M, Buske FA, Frith M, Grant CE, Clementi L, Ren J, Li WW, Noble WS.</td></tr>
<tr class='citation'><td class='top'>[16]</td><td class='left' >R: A language and environment for statistical computing; R Core Team. R Foundation for Statistical Computing, Vienna, Austria, (2013)</td></tr>
</tbody></table>
</div></div>
<hr><span><img src='includes/images/favicon-32x32.png' /> Report generated with NGSANE version: 954f6ef71cd01bfe8330e46089925ade0c2afc94 (git hash)</span><span style='float:right;'>Last modified: Wed Sep 10 14:58:57 EST 2014</span>
</div><!-- center --></body>

<script type=text/javascript>
    var tables = [];
    $(document).ready(function() { 
        for (var i = 0; i < datatable_array.length; i++) { 
            eval("var "+datatable_array[i].html+" = "+datatable_array[i].json);
            eval("tables.push($(\"#"+datatable_array[i].html+"\").dataTable("+datatable_array[i].json+"));");
        }
        $("#search").keyup(function(){
            for (var i=0;i<tables.length;i++){
                tables[i].fnDraw();
            }
        });
        $("#showAggregation").click(function(){
            for (var i=0;i<tables.length;i++){
                tables[i].fnDraw();
            }
        });
    }); 

//This script onclick shows different tabs
$( "div.panel div.tabs a" ).click(function() {

    //Finds old selected tablink (heading) and sets to unselected
//    console.log($(this).parent().find("a.selected"));
    $(this).parent().find("a.selected").attr('class', "");
    
    //Sets new tab heading to selected
//    console.log($(this).attr('href'));
    $(this).attr('class', "selected");
//    console.log($(this));
    
    //Using id from tab heading find content div id
    var id = getHash ($(this).attr('href'));
    var ContentDivId = "DC_"+id;
    //console.log(ContentDivId);
    
    //Sets old content div to tabContent hide
//    console.log($('#' + ContentDivId).parent().find("div.tabContent"));
    $('#' + ContentDivId).parent().find("div.tabContent").attr('class', "tabContent_hide");
    
    //Sets new content div to tabContent (able to be seen)
    //console.log($('#' + ContentDivId).attr('class'));
    $('#' + ContentDivId).attr('class', "tabContent");
    //console.log($('#' + ContentDivId).attr('class'));

});

// add filtering for invalid rows
$.fn.dataTableExt.afnFiltering.push(
	 function( oSettings, aData, iDataIndex ) {
		
		var search = $("#search");
		if ($("#search")[0].value=="") {
             console.log("No search value");
			return true;
		} else {
			var searchArray = $("#search")[0].value.split(/\s+/);
			console.log(searchArray);
			var sArrayLength = searchArray.length;
			var numColumns = oSettings.aoColumns.length;
			//check relevant columns using 
			// cycle through columns
			var display = 0;
			//document.write(searchArray[0]);
			
			//document.write(sArrayLength);
			//document.write(numColumns);
			var iColumn = 0;
			while (iColumn < numColumns && display == 0) {
						
				//go through search array and check if cell regex matches any search items..
				for (isTerm = 0; isTerm < sArrayLength; isTerm++) {
					//alert (
					//document.write(searchArray[isTerm].length);
					if (searchArray[isTerm].length != 0) {
						var patt1=new RegExp(searchArray[isTerm]);
						//document.write(patt1);
						if(patt1.test(aData[iColumn])) {
							display = 1;
						}
					}

				}
				iColumn++;
			}
			if (display == 1) {
				return true;
			} else {
				return false;
			}
		}
	}
);
</script>

