
<!DOCTYPE html><html>
<head><meta charset="windows-1252"><title>NGSANE project card</title>
<script type="text/javascript" src="includes/js/jquery.js"></script>
<script type="text/javascript" charset="utf8" src="includes/js/jquery.dataTables.min.js"></script>
<script type="text/javascript" charset="utf8" src="includes/js/dataTables.responsive.js"></script>
<link rel="stylesheet" type="text/css" href="includes/css/jquery.dataTables.min.css">
<link rel="stylesheet" type="text/css" href="includes/css/dataTables.responsive.css">
<link rel="stylesheet" type="text/css" href="includes/css/ngsane.css">
<script type=text/javascript src=includes/js/ngsane.js></script>
<link rel="shortcut icon" href="includes/images/favicon.ico">
<link rel="apple-touch-icon" sizes="57x57" href="includes/images/apple-touch-icon-57x57.png">
<link rel="apple-touch-icon" sizes="114x114" href="includes/images/apple-touch-icon-114x114.png">
<link rel="apple-touch-icon" sizes="72x72" href="includes/images/apple-touch-icon-72x72.png">
<link rel="apple-touch-icon" sizes="144x144" href="includes/images/apple-touch-icon-144x144.png">
<link rel="apple-touch-icon" sizes="60x60" href="includes/images/apple-touch-icon-60x60.png">
<link rel="apple-touch-icon" sizes="120x120" href="includes/images/apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" sizes="76x76" href="includes/images/apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" sizes="152x152" href="includes/images/apple-touch-icon-152x152.png">
<link rel="icon" type="image/png" href="includes/images/favicon-196x196.png" sizes="196x196">
<link rel="icon" type="image/png" href="includes/images/favicon-160x160.png" sizes="160x160">
<link rel="icon" type="image/png" href="includes/images/favicon-96x96.png" sizes="96x96">
<link rel="icon" type="image/png" href="includes/images/favicon-16x16.png" sizes="16x16">
<link rel="icon" type="image/png" href="includes/images/favicon-32x32.png" sizes="32x32">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="msapplication-TileImage" content="includes/images/mstile-144x144.png">
<meta name="msapplication-config" content="includes/images/browserconfig.xml">
</head><body>
<script type="text/javascript">
//<![CDATA[
    var datatable_array=[];
//]]>
</script>

<div id="center">
<div class=panel id=quicklinks><h2>Quicklinks</h2><div>

<a href='#bwa_panel' class='quicklinks'>bwa</a>|<a href='#reCalAln_panel' class='quicklinks'>reCalAln</a>|<a href='#variant_panel' class='quicklinks'>variant</a>|<a href='#bamann_bwa_panel' class='quicklinks'>bamann_bwa</a>|<a href='#pindel_panel' class='quicklinks'>pindel</a>|<a href='#References_panel' class='quicklinks'>References</a>

<div id ="right"><label>Aggregate: <input id="showAggregation" type="checkbox" /></label> <label>Global Filter: <input type="search" id="search" /></label></div>
</div><!-- Links -->
</div><!-- panel -->
<div class='panel' id='bwa_panel'><div class='tabs headbagb' id='bwa_panel_tabs'><a name='bwa'></a>
            <a class='selected' href='#bwa_results'><h2 id='bwa_h_results' class='sub'>BWA</h2></a>
            <a href='#bwa_checklist'><h2 id='bwa_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='bwa_counter_checkpoints_passed'></span><span class='failed' id='bwa_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#bwa_notes'><h2 id='bwa_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='bwa_counter_notes'></span></span></span></h2></a>
            <a href='#bwa_errors'><h2 id='bwa_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='bwa_counter_errors'></span></span></h2></a>
            <a href='#bwa_logfiles'><h2 id='bwa_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
<a href='#bwa_PRIMARY_RESULT_FILES'><h2 id='bwa_h_nrfiles' class='sub inactive' rel='files'>Primary result files</h2></a>
</div>
<div class='tabContent_hide' id='DC_bwa_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/bwa 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 2/2 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/2 have # QCVARIABLES
QC_PASS .. 0/2 have We are loosing reads
QC_PASS .. 0/2 have MAPQ should be 0 for unmapped read
QC_PASS .. 0/2 have no such file
QC_PASS .. 0/2 have file not found
QC_PASS .. 0/2 have bwa.sh: line
QC_PASS .. 0/2 have Resource temporarily unavailable
>>>>>>>>>> CheckPoints 
QC_PASS .. 2/2 have programs
QC_PASS .. 2/2 have parameters
QC_PASS .. 2/2 have recall files from tape
QC_PASS .. 2/2 have BWA
QC_PASS .. 2/2 have bam sorting
QC_PASS .. 2/2 have clean sam
QC_PASS .. 2/2 have mark duplicates
QC_PASS .. 2/2 have statistics
QC_PASS .. 2/2 have calculate inner distance
QC_PASS .. 2/2 have samstat
QC_PASS .. 2/2 have verify
QC_PASS .. 2/2 have cleanup

</pre></div></div>
<div class='tabContent_hide' id='DC_bwa_notes'><div class='box scroll'><pre>

qout/bwa/Transcript_RNA3kChr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bwa/job.20140910.log
[NOTE] set java parameters
[NOTE] EXPID OmicsRNA; LIBRARY AGRF; PLATFORM Illumina
[NOTE] BWA algorithm is sampe
[NOTE] Detect fastq Phred encoding
[NOTE] Illumina 1.8 Phred33	(41, 74) fastq format detected
[NOTE] PAIRED READS
[NOTE] fixmate
[NOTE] PASS check mapping: 7030 == 7030

qout/bwa/Transcript_RNA3kChr16rep.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bwa/job.20140910.log
[NOTE] set java parameters
[NOTE] EXPID OmicsRNA; LIBRARY AGRF; PLATFORM Illumina
[NOTE] BWA algorithm is sampe
[NOTE] Detect fastq Phred encoding
[NOTE] Illumina 1.8 Phred33	(41, 74) fastq format detected
[NOTE] PAIRED READS
[NOTE] fixmate
[NOTE] PASS check mapping: 7030 == 7030
</pre></div></div>
<div class='tabContent_hide' id='DC_bwa_errors'><div class='box scroll'><pre>

qout/bwa/Transcript_RNA3kChr16.out
-- all good, no errors

qout/bwa/Transcript_RNA3kChr16rep.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_bwa_logfiles'><div><div class='box scroll'>
<a href='../qout/bwa/Transcript_RNA3kChr16.out'>Transcript_RNA3kChr16.out</a><br/>
<a href='../qout/bwa/Transcript_RNA3kChr16rep.out'>Transcript_RNA3kChr16rep.out</a><br/>
</div></div></div>
<div class='tabContent_hide' id='DC_bwa_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
<a href='../Transcript/bwa/RNA3kChr16.asd.bam'>RNA3kChr16.asd.bam</a><br/>
<a href='../Transcript/bwa/RNA3kChr16rep.asd.bam'>RNA3kChr16rep.asd.bam</a><br/>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#bwa_counter_notes').text('18');
            $('#bwa_counter_errors').text('0');
            $('#bwa_counter_checkpoints_passed').text('20');
            $('#bwa_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#bwa_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#bwa_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#bwa_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#bwa_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_bwa_results'><div>
<script type="text/javascript">
//<![CDATA[
bwa_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Total reads', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Qfail', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Duplicates', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'%', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'Mapped', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'%', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'Mapped pairs', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'%', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'Singletons', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
['Transcript' , '<a target="detail" href="../Transcript/bwa/RNA3kChr16.asd.bam.html">RNA3kChr16</a>' , 7030.00 , 0.00 , 433.00 , 6.15 , 5638.00 , 80.19 , 2615.00 , 74.39 , 354.00 ]
,['Transcript' , '<a target="detail" href="../Transcript/bwa/RNA3kChr16rep.asd.bam.html">RNA3kChr16rep</a>' , 7030.00 , 0.00 , 433.00 , 6.15 , 5638.00 , 80.19 , 2615.00 , 74.39 , 354.00 ]
        ]};
        datatable_array.push({"json":"bwa_Table_1_json","html":"bwa_id_Table_1"});
        //]]>
        </script>
    <table id="bwa_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel' id='reCalAln_panel'><div class='tabs headbagb' id='reCalAln_panel_tabs'><a name='reCalAln'></a>
            <a class='selected' href='#reCalAln_results'><h2 id='reCalAln_h_results' class='sub'>Recalibrate + Realign</h2></a>
            <a href='#reCalAln_checklist'><h2 id='reCalAln_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='reCalAln_counter_checkpoints_passed'></span><span class='failed' id='reCalAln_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#reCalAln_notes'><h2 id='reCalAln_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='reCalAln_counter_notes'></span></span></span></h2></a>
            <a href='#reCalAln_errors'><h2 id='reCalAln_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='reCalAln_counter_errors'></span></span></h2></a>
            <a href='#reCalAln_logfiles'><h2 id='reCalAln_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
</div>
<div class='tabContent_hide' id='DC_reCalAln_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/reCalAln 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 2/2 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/2 have # QCVARIABLES
QC_PASS .. 0/2 have We are loosing reads
QC_PASS .. 0/2 have for unmapped read
QC_PASS .. 0/2 have no such file
QC_PASS .. 0/2 have file not found
QC_PASS .. 0/2 have reCalAln.sh: line
>>>>>>>>>> CheckPoints 
QC_PASS .. 2/2 have programs
QC_PASS .. 2/2 have parameters
QC_PASS .. 2/2 have recall files from tape
QC_PASS .. 2/2 have find intervals to improve
QC_PASS .. 2/2 have realign
QC_PASS .. 2/2 have count covariantes 
QC_PASS .. 2/2 have adjust score
QC_PASS .. 2/2 have evaluate performace
QC_PASS .. 2/2 have count covariantes after recalibration
QC_PASS .. 2/2 have sort/index
QC_PASS .. 2/2 have statistics

</pre></div></div>
<div class='tabContent_hide' id='DC_reCalAln_notes'><div class='box scroll'><pre>

qout/reCalAln/Transcript_RNA3kChr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/reCalAln/job.20140910.log
[NOTE] set java parameters
[NOTE] PAIRED
[NOTE] GATK multithreading
[NOTE] fixmate
[NOTE] PASS check recalibration and realignment: 7030 == 7030

qout/reCalAln/Transcript_RNA3kChr16rep.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/reCalAln/job.20140910.log
[NOTE] set java parameters
[NOTE] PAIRED
[NOTE] GATK multithreading
[NOTE] fixmate
[NOTE] PASS check recalibration and realignment: 7030 == 7030
</pre></div></div>
<div class='tabContent_hide' id='DC_reCalAln_errors'><div class='box scroll'><pre>

qout/reCalAln/Transcript_RNA3kChr16.out
-- all good, no errors

qout/reCalAln/Transcript_RNA3kChr16rep.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_reCalAln_logfiles'><div><div class='box scroll'>
<a href='../qout/reCalAln/Transcript_RNA3kChr16.out'>Transcript_RNA3kChr16.out</a><br/>
<a href='../qout/reCalAln/Transcript_RNA3kChr16rep.out'>Transcript_RNA3kChr16rep.out</a><br/>
</div></div></div>
<div class='tabContent_hide' id='DC_reCalAln_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
<a href='../Transcript/reCalAln/RNA3kChr16.asd.bam'>RNA3kChr16.asd.bam</a><br/>
<a href='../Transcript/reCalAln/RNA3kChr16rep.asd.bam'>RNA3kChr16rep.asd.bam</a><br/>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#reCalAln_counter_notes').text('12');
            $('#reCalAln_counter_errors').text('0');
            $('#reCalAln_counter_checkpoints_passed').text('18');
            $('#reCalAln_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#reCalAln_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#reCalAln_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#reCalAln_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#reCalAln_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_reCalAln_results'><div>
<script type="text/javascript">
//<![CDATA[
reCalAln_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Total reads', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Qfail', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Duplicates', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'%', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'Mapped', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'%', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'Mapped pairs', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'%', 'mRender': function(data){return formatNumber(data);} , 'sWidth':'40px'} ,{'sTitle':'Singletons', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
['Transcript' , 'RNA3kChr16' , 7030.00 , 0.00 , 433.00 , 6.15 , 5638.00 , 80.19 , 2615.00 , 74.39 , 354.00 ]
,['Transcript' , 'RNA3kChr16rep' , 7030.00 , 0.00 , 433.00 , 6.15 , 5638.00 , 80.19 , 2615.00 , 74.39 , 354.00 ]
        ]};
        datatable_array.push({"json":"reCalAln_Table_1_json","html":"reCalAln_id_Table_1"});
        //]]>
        </script>
    <table id="reCalAln_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel' id='variant_panel'><div class='tabs headbagb' id='variant_panel_tabs'><a name='variant'></a>
            <a class='selected' href='#variant_results'><h2 id='variant_h_results' class='sub'>VarCalls</h2></a>
            <a href='#variant_checklist'><h2 id='variant_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='variant_counter_checkpoints_passed'></span><span class='failed' id='variant_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#variant_notes'><h2 id='variant_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='variant_counter_notes'></span></span></span></h2></a>
            <a href='#variant_errors'><h2 id='variant_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='variant_counter_errors'></span></span></h2></a>
            <a href='#variant_logfiles'><h2 id='variant_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
</div>
<div class='tabContent_hide' id='DC_variant_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/gatkVARs 
###################################################
>>>>>>>>>> Finished
**_FAIL .. 0/1 are finished
>>>>>>>>>> Errors
>>>>>>>>>> CheckPoints 
QC_PASS .. 1/1 have programs
QC_PASS .. 1/1 have parameters
QC_PASS .. 1/1 have recall files from tape
QC_PASS .. 1/1 have call snps
QC_PASS .. 1/1 have hardfilter
QC_PASS .. 1/1 have index for IGV
QC_PASS .. 1/1 have evaluate
**_FAIL .. 0/1 have re-calibrate

</pre></div></div>
<div class='tabContent_hide' id='DC_variant_notes'><div class='box scroll'><pre>
</pre></div></div>
<div class='tabContent_hide' id='DC_variant_errors'><div class='box scroll'><pre>
</pre></div></div>
<div class='tabContent_hide' id='DC_variant_logfiles'><div><div class='box scroll'>
</div></div></div>
<div class='tabContent_hide' id='DC_variant_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#variant_counter_notes').text('0');
            $('#variant_counter_errors').text('0');
            $('#variant_counter_checkpoints_passed').text('7');
            $('#variant_counter_checkpoints_failed').text('2'); 
            if (0==0){
                $('#variant_counter_errors').toggleClass('errors neutral');
            }; 
            if(2==0){
                $('#variant_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#variant_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#variant_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_variant_results'><div>
<script type="text/javascript">
//<![CDATA[
variant_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Total', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Known', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'SNPdb Conc', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Variant Rate/Bp', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Het/Hom', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Novel', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Variant Rate/Bp', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Het/Hom', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
['Transcript' , 'Transcript' , 92.00 , 44.00 , 97.73 , 2053517.00 , 0.33 , 48.00 , 1882390.00 , 0.07 ]
        ]};
        datatable_array.push({"json":"variant_Table_1_json","html":"variant_id_Table_1"});
        //]]>
        </script>
    <table id="variant_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
<script type="text/javascript">
//<![CDATA[
variant_Table_2_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Total', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Known', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'SNPdb Conc', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Variant Rate/Bp', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Het/Hom', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Novel', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Variant Rate/Bp', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Het/Hom', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
        ]};
        datatable_array.push({"json":"variant_Table_2_json","html":"variant_id_Table_2"});
        //]]>
        </script>
<hr class='spacer' />
    <table id="variant_id_Table_2" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
<script type="text/javascript">
//<![CDATA[
variant_Table_3_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Total', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Known', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'SNPdb Conc', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Variant Rate/Bp', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Het/Hom', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Novel', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Variant Rate/Bp', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Het/Hom', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
['Transcript' , 'Transcript' , 5.00 , 0.00 , 0.00 , 0.00 , 0.00 , 5.00 , 18070950.00 , 0.67 ]
        ]};
        datatable_array.push({"json":"variant_Table_3_json","html":"variant_id_Table_3"});
        //]]>
        </script>
<hr class='spacer' />
    <table id="variant_id_Table_3" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel' id='bamann_bwa_panel'><div class='tabs headbagb' id='bamann_bwa_panel_tabs'><a name='bamann_bwa'></a>
            <a class='selected' href='#bamann_bwa_results'><h2 id='bamann_bwa_h_results' class='sub'>AnnotateBam</h2></a>
            <a href='#bamann_bwa_checklist'><h2 id='bamann_bwa_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='bamann_bwa_counter_checkpoints_passed'></span><span class='failed' id='bamann_bwa_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#bamann_bwa_notes'><h2 id='bamann_bwa_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='bamann_bwa_counter_notes'></span></span></span></h2></a>
            <a href='#bamann_bwa_errors'><h2 id='bamann_bwa_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='bamann_bwa_counter_errors'></span></span></h2></a>
            <a href='#bamann_bwa_logfiles'><h2 id='bamann_bwa_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
</div>
<div class='tabContent_hide' id='DC_bamann_bwa_checklist'><div><pre>
###################################################
# NGSANE /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/mods/annotateBam 
###################################################
>>>>>>>>>> Finished
QC_PASS .. 2/2 are finished
>>>>>>>>>> Errors
QC_PASS .. 0/2 have # QCVARIABLES
QC_PASS .. 0/2 have Resource temporarily unavailable
>>>>>>>>>> CheckPoints 
QC_PASS .. 2/2 have programs
QC_PASS .. 2/2 have parameters
QC_PASS .. 2/2 have recall files from tape
QC_PASS .. 2/2 have bedmerge
QC_PASS .. 2/2 have run annotation
QC_PASS .. 2/2 have summarize

</pre></div></div>
<div class='tabContent_hide' id='DC_bamann_bwa_notes'><div class='box scroll'><pre>

qout/bamann_bwa/Transcript_RNA3kChr16.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bamann_bwa/job.20140910.log
[NOTE] using libraries in /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann/
[NOTE] annotate with  exon intron lincRNA small_ncRNA tRNA UCSC_rRNA  /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//exon.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//intron.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//lincRNA.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//small_ncRNA.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//tRNA.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//UCSC_rRNA.gtf

qout/bamann_bwa/Transcript_RNA3kChr16rep.out
[NOTE] Jobfile: /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/qout/bamann_bwa/job.20140910.log
[NOTE] using libraries in /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann/
[NOTE] annotate with  exon intron lincRNA small_ncRNA tRNA UCSC_rRNA  /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//exon.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//intron.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//lincRNA.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//small_ncRNA.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//tRNA.gtf /share/ClusterShare/software/contrib/fabbus/ngsane/dev1/smokebox/referenceData/bamann//UCSC_rRNA.gtf
</pre></div></div>
<div class='tabContent_hide' id='DC_bamann_bwa_errors'><div class='box scroll'><pre>

qout/bamann_bwa/Transcript_RNA3kChr16.out
-- all good, no errors

qout/bamann_bwa/Transcript_RNA3kChr16rep.out
-- all good, no errors
</pre></div></div>
<div class='tabContent_hide' id='DC_bamann_bwa_logfiles'><div><div class='box scroll'>
<a href='../qout/bamann_bwa/Transcript_RNA3kChr16.out'>Transcript_RNA3kChr16.out</a><br/>
<a href='../qout/bamann_bwa/Transcript_RNA3kChr16rep.out'>Transcript_RNA3kChr16rep.out</a><br/>
</div></div></div>
<div class='tabContent_hide' id='DC_bamann_bwa_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#bamann_bwa_counter_notes').text('6');
            $('#bamann_bwa_counter_errors').text('0');
            $('#bamann_bwa_counter_checkpoints_passed').text('9');
            $('#bamann_bwa_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#bamann_bwa_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#bamann_bwa_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#bamann_bwa_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#bamann_bwa_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_bamann_bwa_results'><div>
</div></div></div>
<div class='panel' id='pindel_panel'><div class='tabs headbagb' id='pindel_panel_tabs'><a name='pindel'></a>
            <a class='selected' href='#pindel_results'><h2 id='pindel_h_results' class='sub'>Structural Variants (Pindel)</h2></a>
            <a href='#pindel_checklist'><h2 id='pindel_h_checklist' class='sub inactive' rel='checklist'>Checklist<span class='counter'><span class='passed' id='pindel_counter_checkpoints_passed'></span><span class='failed' id='pindel_counter_checkpoints_failed'></span></span></h2></a>
            <a href='#pindel_notes'><h2 id='pindel_h_notes' class='sub inactive' rel='notes'>Notes<span class='counter'><span class='neutral' id='pindel_counter_notes'></span></span></span></h2></a>
            <a href='#pindel_errors'><h2 id='pindel_h_errors' class='sub inactive' rel='notes'>Errors<span class='counter'><span class='errors' id='pindel_counter_errors'></span></span></h2></a>
            <a href='#pindel_logfiles'><h2 id='pindel_h_logfiles' class='sub inactive' rel='errors'>Log files</h2></a>
</div>
<div class='tabContent_hide' id='DC_pindel_checklist'><div><pre>
</pre></div></div>
<div class='tabContent_hide' id='DC_pindel_notes'><div class='box scroll'><pre>
</pre></div></div>
<div class='tabContent_hide' id='DC_pindel_errors'><div class='box scroll'><pre>
</pre></div></div>
<div class='tabContent_hide' id='DC_pindel_logfiles'><div><div class='box scroll'>
</div></div></div>
<div class='tabContent_hide' id='DC_pindel_PRIMARY_RESULT_FILES'><div><div class='box scroll'>
</div></div></div>
<script type='text/javascript'> 
        if (typeof jQuery === 'undefined') {
            console.log('jquery not loaded');
        } else {
            $('#pindel_counter_notes').text('0');
            $('#pindel_counter_errors').text('0');
            $('#pindel_counter_checkpoints_passed').text('0');
            $('#pindel_counter_checkpoints_failed').text('0'); 
            if (0==0){
                $('#pindel_counter_errors').toggleClass('errors neutral');
            }; 
            if(0==0){
                $('#pindel_counter_checkpoints_failed').toggleClass('failed nofailed');
            };

            $('#pindel_panelback h2').click(function() {
                $(this).parent().children().addClass('inactive');
                $(this).removeClass('inactive');
                console.log();
                var panel=$(this).attr('id').replace('_h_','_');
                $('#pindel_panel div.wrapper div.display').html($('#'+ panel).html());
            });
        }
    </script>
<div class='tabContent' id='DC_pindel_results'><div>
<script type="text/javascript">
//<![CDATA[
pindel_Table_1_json={
	"bPaginate": true,
	"bProcessing": false,
	"bAutoWidth": false,
	"bInfo": true,
	"bLengthChange": true,
    "bFilter": true,
	"iDisplayLength": 10,
	"bJQueryUI": false,
	"aLengthMenu": [[1, 5, 10, -1], [1, 5, 10, "All"]],
    "footerCallback": function ( row, data, start, end, display ) {
        var api = this.api(), data;
        // empty footer
        if (! $("#showAggregation")[0].checked){
            for (var c=0; c<this.fnSettings().aoColumns.length; c++){
                $( api.column( c ).footer() ).html("");
            }
        } else if (data.length <= 1){
            $( api.column( 1 ).footer() ).html("insufficient data for aggregation");
        } else {
        
            $( api.column( 1 ).footer() ).html("sum</br>min<br>max<br>median</br>mean</br>stdev");
            
            for (var c=2; c<this.fnSettings().aoColumns.length; c++){
    
                // Remove the formatting to get integer data for summation
                var intVal = function ( i ) {
                    var f = parseFloat(i);
                    if (!isNaN(f))
                        return f;
                    else
                        return "";
                };
        
                // Total over all pages
                var data = api.column( c, {"filter":"applied"} ).data();
                var total = data.length ? data.reduce( function (a, b) {  return intVal(a) + intVal(b); } ) : "";
                var maxV = Math.max.apply(Math, data);
                var minV = Math.min.apply(Math, data);
                var median = Math.median.apply(Math,data);
                var mean = Math.mean.apply(Math,data);
                var stdev = Math.std.apply(Math,data);
        
                $( api.column( c ).footer() ).html(formatNumber(total)+"<br/>"+formatNumber(minV)+"<br/>"+formatNumber(maxV)+"<br/>"+formatNumber(median)+"<br/>"+formatNumber(mean)+"<br/>"+formatNumber(stdev)+"<br/>");
            }
        }
    },
	"aoColumns": [

{'sTitle':'Experiment', 'sClass':'left'}, {'sTitle':'Sample', 'sClass':'left'} ,{'sTitle':'Total', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Known', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'SNPdb Conc', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Variant Rate/Bp', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Het/Hom', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Novel', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Variant Rate/Bp', 'mRender': function(data){return formatNumber(data);} } ,{'sTitle':'Het/Hom', 'mRender': function(data){return formatNumber(data);} }
], "aaData": [
['Transcript' , 'joined.vcf' , 16.00 , 0.00 , 0.00 , 0.00 , 0.00 , 11.00 , 8214068.00 , 0.00 ]
        ]};
        datatable_array.push({"json":"pindel_Table_1_json","html":"pindel_id_Table_1"});
        //]]>
        </script>
    <table id="pindel_id_Table_1" class="display no-wrap" width="100%"><thead><tr><th></th><th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th> <th></th></tr></thead><tbody><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tbody><tfoot><tr><td></td><td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td> <td></td></tr></tfoot></table>
</div></div></div>
<div class='panel'><div class='headbagb'><a name='References_panel'><h2 class='sub'>References</h2></a></div>
<h3>Please cite the following publications/resources</h3>
<table class="data"><thead><tr><th><div style="width:10px"></div></th><th><div></div></th></tr></thead><tbody>
<tr class='citation'><td class='top'>[1]</td><td class='left' >Bioinformatics. 2009 Aug 15;25(16):2078-9. doi: 10.1093/bioinformatics/btp352. Epub 2009 Jun 8. 'The Sequence Alignment/Map format and SAMtools.'; Li H, Handsaker B, Wysoker A, Fennell T, Ruan J, Homer N, Marth G, Abecasis G, Durbin R; 1000 Genome Project Data Processing Subgroup.</td></tr>
<tr class='citation'><td class='top'>[2]</td><td class='left' >Bioinformatics. 2010 Mar 1;26(5):589-95. doi: 10.1093/bioinformatics/btp698. Epub 2010 Jan 15. 'Fast and accurate long-read alignment with Burrows-Wheeler transform.'; Li H, Durbin R.</td></tr>
<tr class='citation'><td class='top'>[3]</td><td class='left' >Bioinformatics. 2010 Mar 15;26(6):841-2. doi: 10.1093/bioinformatics/btq033. Epub 2010 Jan 28. 'BEDTools: a flexible suite of utilities for comparing genomic features.; Quinlan AR, Hall IM.</td></tr>
<tr class='citation'><td class='top'>[4]</td><td class='left' >Bioinformatics. 2011 Jan 1;27(1):130-1. doi: 10.1093/bioinformatics/btq614. Epub 2010 Nov 18. 'SAMStat: monitoring biases in next generation sequencing data.'; Lassmann T, Hayashizaki Y, Daub CO.</td></tr>
<tr class='citation'><td class='top'>[5]</td><td class='left' >Bioinformatics 2014; doi: 10.1093/bioinformatics/btu036 'NGSANE: A Lightweight Production Informatics Framework for High Throughput Data Analysis; Buske FA, French HJ, Smith MA, Clark SJ, Bauer DC</td></tr>
<tr class='citation'><td class='top'>[6]</td><td class='left' >Brief Bioinform. 2013 Mar;14(2):144-61. doi: 10.1093/bib/bbs038. Epub 2012 Aug 20. 'The UCSC genome browser and associated tools.'; Kuhn RM, Haussler D, Kent WJ.</td></tr>
<tr class='citation'><td class='top'>[7]</td><td class='left' >Brief Bioinform. 2013 Mar;14(2):178-92. doi: 10.1093/bib/bbs017. Epub 2012 Apr 19. 'Integrative Genomics Viewer (IGV): high-performance genomics data visualization and exploration.'; Thorvaldsdóttir H, Robinson JT, Mesirov JP.</td></tr>
<tr class='citation'><td class='top'>[8]</td><td class='left' >Genome Res. 2010 Sep;20(9):1297-303. doi: 10.1101/gr.107524.110. Epub 2010 Jul 19. 'The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data. McKenna A, Hanna M, Banks E, Sivachenko A, Cibulskis K, Kernytsky A, Garimella K, Altshuler D, Gabriel S, Daly M, DePristo MA.</td></tr>
<tr class='citation'><td class='top'>[9]</td><td class='left' >http://picard.sourceforge.net.</td></tr>
<tr class='citation'><td class='top'>[10]</td><td class='left' >Nucleic Acids Res. 2012 Jul;40(Web Server issue):W622-7. doi: 10.1093/nar/gks540. Epub 2012 Jun 8. 'RobiNA: a user-friendly, integrated software solution for RNA-Seq-based transcriptomics.'; Lohse M, Bolger AM, Nagel A, Fernie AR, Lunn JE, Stitt M, Usadel B.</td></tr>
<tr class='citation'><td class='top'>[11]</td><td class='left' >R: A language and environment for statistical computing; R Core Team. R Foundation for Statistical Computing, Vienna, Austria, (2013)</td></tr>
</tbody></table>
</div></div>
<hr><span><img src='includes/images/favicon-32x32.png' /> Report generated with NGSANE version: da79b398b0186a1acd1ffedc36ead4cdddc71955 (git hash)</span><span style='float:right;'>Last modified: Wed Sep 10 11:01:00 EST 2014</span>
</div><!-- center --></body>

<script type=text/javascript>
    var tables = [];
    $(document).ready(function() { 
        for (var i = 0; i < datatable_array.length; i++) { 
            eval("var "+datatable_array[i].html+" = "+datatable_array[i].json);
            eval("tables.push($(\"#"+datatable_array[i].html+"\").dataTable("+datatable_array[i].json+"));");
        }
        $("#search").keyup(function(){
            for (var i=0;i<tables.length;i++){
                tables[i].fnDraw();
            }
        });
        $("#showAggregation").click(function(){
            for (var i=0;i<tables.length;i++){
                tables[i].fnDraw();
            }
        });
    }); 
</script>

<script type='text/javascript' src='includes/js/genericJavaScript.js'></script></html>
